### PR TITLE
Update webpack-cli

### PR DESCRIPTION
**Pull request by JonWallsten**
Implemented compilation.emitAssets for additional resources
Replaced afterProcessAssets with processAssets hook

### CODE REVIEW

File: src/instances.ts [hunk 1]
```diff
@@ -316,7 +316,7 @@ function addAssetHooks(loader: WebpackLoaderContext, instance: TSInstance) {
   // compilation is actually of type webpack.Compilation, but afterProcessAssets
   // only exists in webpack5 and at the time of writing ts-loader is built using webpack4
   const makeAssetsCallback = (compilation: any) => {
-    compilation.hooks.afterProcessAssets.tap('ts-loader', () =>
+    compilation.hooks.processAssets.tap({ name: 'ts-loader', stage: webpack.Compilation.PROCESS_ASSETS_STAGE_ADDITIONAL }, (_assets: webpack.Asset[]) => {
```
Comment (JonWallsten) [inline src/instances.ts:319]: I'm not sure what we decided in this one? It's the same thing as the fix above. Both works, but one is newer and preferred.
`_assets: webpack.Asset[]` could be removed if you prefer leaving info out rather then keep it for future usage. I usually leave it so you don't have to dig it up when you need it. But I don't have a strong opinion about it.

**Comment by johnnyreilly** [comment]
Cool - this looks great - thanks!

I'd like to merge this, but in terms of sequencing, I'd like to get the mystery of the missing `bundle.js` files solved with the comparison test pack first. Then we're well placed to see the difference that taking this PR makes to output / behaviour.

**Comment by JonWallsten** [comment]
> Cool - this looks great - thanks!
> 
> I'd like to merge this, but in terms of sequencing, I'd like to get the mystery of the missing `bundle.js` files solved with the comparison test pack first. Then we're well placed to see the difference that taking this PR makes to output / behaviour.

Yeah, I was thinking the same! I'll help out once I'm back home!

**Comment by JonWallsten** [comment]
Hmm, I'm trying this out with a symlinked ts-loader that  i just built, and it works fine for me:
image

image

**Comment by JonWallsten** [comment]
@johnnyreilly: What's going on here?
image

**Comment by johnnyreilly** [comment]
I think I've solved it - gimme 10 mins and I'm going to push on the other branch

**Comment by johnnyreilly** [comment]
You probably want to revert your test data commit - there's an issue with it that my fix will solve

**Comment by johnnyreilly** [comment]
Pushed

**Review by JonWallsten — COMMENTED** [review]
> I think I've solved it - gimme 10 mins and I'm going to push on the other branch

What was the fix? 😎 I'll do a git reset and rebase on your new changes!

**Comment by johnnyreilly** [comment]
It was this line: https://github.com/TypeStrong/ts-loader/blob/48189c55fe2e5c40ed2453b54f02343d9536a60b/test/comparison-tests/create-and-execute-test.js#L246

**Comment by JonWallsten** [comment]
@johnnyreilly: Just rebased to trigger a new test run. Same issue as yesterday in you branch.
That path looks really weird:

```
Error: EINVAL: invalid argument, mkdir 'C:\Users\RUNNER~1\AppData\Local\Temp\_karma_webpack_313335\D:\a\ts-loader\ts-loader\test\execution-tests\3.0.1_projectReferences\lib'
```

**Comment by johnnyreilly** [comment]
Yeah - really does

**Comment by JonWallsten** [comment]
> Yeah - really does

@johnnyreilly: I'm not sure how to solve this.
Somewhere someone is joining `require(os).tempDir()` and the `projectReference.path`, and since they are on different drives I guess they URL end's up wrong. I can't reproduce this locally though. Maybe differences in node or Windows version.

**Comment by JonWallsten** [comment]
It also seems to be a race condition:
image
image
Mostly all of them fails, but in two cases one of the test cases passes.

**Comment by JonWallsten** [comment]
And why is the project reference an absolute path? 
Seems weird when the path is:
image

**Comment by JonWallsten** [comment]
@alexander-akait: I'm having issues with these failing tests, only on Windows and only the GitHub Workflow CI, I can't reproduce this locally. The error happens during Webpack compilation, but is not necessarily Webpacks fault.

`karma-webpack` creates a new `Webpack` instance:
image
`output.path` is set to `"C:\Users\***\AppData\Local\Temp\_karma_webpack_220434"`.
The files are located in `D:/ts-loader/test/execution-tests/3.0.1_projectReferences/`. 
The issue is not that the files are being located on the `D:\` drive itself, because all test cases without TypeScript Project References works fine. So something is special about the files located in the Project reference.

In the end we get this error message: 

> Error: EINVAL: invalid argument, mkdir 'C:\Users\RUNNER~1\AppData\Local\Temp\_karma_webpack_313335\D:\a\ts-loader\ts-loader\test\execution-tests\3.0.1_projectReferences\lib'

So something is appending the `Project reference path` to the `output.path` which of course does't work since Windows don't allow you to add more then one ':' in the path. Out of all my test runs these tests has passed two or three times. Which gives me the feeling it's a race condition. I've been digging a full day soon without finding the issue. 
My question to you is: How does `Webpack` in general handle files when they are located on a different drive than the output path? How do you even know what the relative path is? Do you use the relative based on the current working directory?

**Comment by JonWallsten** [comment]
@johnnyreilly: https://docs.github.com/en/actions/managing-workflow-runs/enabling-debug-logging

Would that help with us getting a Stack Trace for the error? I have tried everything... getting a bit frustrated.

**Comment by JonWallsten** [comment]
Finally a clue:
```
at compiler.webpack.run (D:\a\ts-loader\ts-loader\node_modules\karma-webpack\lib\karma-webpack\controller.js:90:21)
    at finalCallback (D:\a\ts-loader\ts-loader\node_modules\webpack\lib\Compiler.js:390:32)
    at emitAssets.err (D:\a\ts-loader\ts-loader\node_modules\webpack\lib\Compiler.js:417:22)
    at asyncLib.forEachLimit.err (D:\a\ts-loader\ts-loader\node_modules\webpack\lib\Compiler.js:802:22)
    at D:\a\ts-loader\ts-loader\node_modules\neo-async\async.js:2830:7
    at done (D:\a\ts-loader\ts-loader\node_modules\neo-async\async.js:3517:9)
    at writeOut (D:\a\ts-loader\ts-loader\node_modules\webpack\lib\Compiler.js:559:23)
    at fs.mkdir.err (D:\a\ts-loader\ts-loader\node_modules\webpack\lib\util\fs.js:228:4)
    at FSReqWrap.args [as oncomplete] (fs.js:140:20)
```

**Comment by JonWallsten** [comment]
Of course the issue is more complex than that.
**Output from running locally with Windows 10:**
```
Output path: C:\Users\jway52\AppData\Local\Temp\_karma_webpack_905571 

START:
Webpack bundling... 
[Compiler.js] Target file: main.2643079663.js 
[Compiler.js] Target file: runtime.js 
[Compiler.js] Target file: commons.js
1. C:\Users\jway52\AppData\Local\Temp\_karma_webpack_905571 
2. runtime.js 
1. C:\Users\jway52\AppData\Local\Temp\_karma_webpack_905571
2. commons.js 
1. C:\Users\jway52\AppData\Local\Temp\_karma_webpack_905571
2. main.2643079663.js
```

**Output from Windows 2019 Server on GitHub Workflows:**
```
Output path: C:\Users\RUNNER~1\AppData\Local\Temp\_karma_webpack_217803

START:
Webpack bundling...
[Compiler.js] Output path:  C:\Users\RUNNER~1\AppData\Local\Temp\_karma_webpack_217803
[Compiler.js] Target file: main.2591016000.js
[Compiler.js] Target file:  D:\a\ts-loader\ts-loader\test\execution-tests\3.0.1_projectReferences\lib\index.js.map
[Compiler.js] Target file: runtime.js
[Compiler.js] Dir:  C:\Users\RUNNER~1\AppData\Local\Temp\_karma_webpack_217803\D:\a\ts-loader\ts-loader\test\execution-tests\3.0.1_projectReferences\lib
[Compiler.js] Target file: commons.js
[Compiler.js] Output path:  C:\Users\RUNNER~1\AppData\Local\Temp\_karma_webpack_217803
[Compiler.js] Target file: D:\a\ts-loader\ts-loader\test\execution-tests\3.0.1_projectReferences\lib\index.js.map
[Compiler.js] Target file:  D:\a\ts-loader\ts-loader\test\execution-tests\3.0.1_projectReferences\lib\index.js
[Compiler.js] Target file: D:\a\ts-loader\ts-loader\test\execution-tests\3.0.1_projectReferences\lib\index.js
[Compiler.js] Dir:  C:\Users\RUNNER~1\AppData\Local\Temp\_karma_webpack_217803\D:\a\ts-loader\ts-loader\test\execution-tests\3.0.1_projectReferences\lib
[Compiler.js] Target file: D:\a\ts-loader\ts-loader\test\execution-tests\3.0.1_projectReferences\lib\index.d.ts
[Compiler.js] Output path:  C:\Users\RUNNER~1\AppData\Local\Temp\_karma_webpack_217803
[Compiler.js] Target file:  D:\a\ts-loader\ts-loader\test\execution-tests\3.0.1_projectReferences\lib\index.d.ts
[Compiler.js] Dir:  C:\Users\RUNNER~1\AppData\Local\Temp\_karma_webpack_217803\D:\a\ts-loader\ts-loader\test\execution-tests\3.0.1_projectReferences\lib
[Compiler.js] Output path:  C:\Users\RUNNER~1\AppData\Local\Temp\_karma_webpack_217803
[Compiler.js] Target file:  D:\a\ts-loader\ts-loader\test\execution-tests\3.0.1_projectReferences\lib\tsconfig.tsbuildinfo
[Compiler.js] Dir:  C:\Users\RUNNER~1\AppData\Local\Temp\_karma_webpack_217803\D:\a\ts-loader\ts-loader\test\execution-tests\3.0.1_projectReferences\lib
{ [Error: EINVAL: invalid argument, mkdir 'C:\Users\RUNNER~1\AppData\Local\Temp\_karma_webpack_217803\D:\a\ts-loader\ts-loader\test\execution-tests\3.0.1_projectReferences\lib']
  errno: -4071,
  code: 'EINVAL',
  syscall: 'mkdir',
  path:
   'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\_karma_webpack_217803\\D:\\a\\ts-loader\\ts-loader\\test\\execution-tests\\3.0.1_projectReferences\\lib' }
Trace
    at compiler.webpack.run (D:\a\ts-loader\ts-loader\node_modules\karma-webpack\lib\karma-webpack\controller.js:87:21)
    at finalCallback (D:\a\ts-loader\ts-loader\node_modules\webpack\lib\Compiler.js:390:32)
    at emitAssets.err (D:\a\ts-loader\ts-loader\node_modules\webpack\lib\Compiler.js:417:22)
    at asyncLib.forEachLimit.err (D:\a\ts-loader\ts-loader\node_modules\webpack\lib\Compiler.js:805:22)
    at D:\a\ts-loader\ts-loader\node_modules\neo-async\async.js:2830:7
    at done (D:\a\ts-loader\ts-loader\node_modules\neo-async\async.js:3517:9)
    at writeOut (D:\a\ts-loader\ts-loader\node_modules\webpack\lib\Compiler.js:559:23)
    at fs.mkdir.err (D:\a\ts-loader\ts-loader\node_modules\webpack\lib\util\fs.js:228:4)
    at FSReqWrap.args [as oncomplete] (fs.js:140:20)
Error: EINVAL: invalid argument, mkdir 'C:\Users\RUNNER~1\AppData\Local\Temp\_karma_webpack_217803\D:\a\ts-loader\ts-loader\test\execution-tests\3.0.1_projectReferences\lib'
[Compiler.js] Target file: D:\a\ts-loader\ts-loader\test\execution-tests\3.0.1_projectReferences\lib\tsconfig.tsbuildinfo
Error: The operation was canceled.
```

**Comment by johnnyreilly** [comment]
No worries @JonWallsten - thanks for your work. Don't feel bad BTW - I've no time and I'm not making progress so I'm pausing for now. I'll look again when I've more time and patience.

**Comment by JonWallsten** [comment]
> No worries @JonWallsten - thanks for your work. Don't feel bad BTW - I've no time and I'm not making progress so I'm pausing for now. I'll look again when I've more time and patience.

Sounds good! Let's take a break. It's so hard to work with these issues without being two people talking and sharing ideas in real time. The most frustrating thing is that since I can't reproduce it locally I have to use forks for each package involved and just read logs. No way of debugging, unless you can actually access the virtual machine we're testing on. Can we do that?

**Comment by johnnyreilly** [comment]
Not to my knowledge, no

**Pull request closed without merging**